Use @fx.wrap for python-side NJT construction

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130505
* __->__ #130183

This PR:
* Changes `nested_view_from_values_offsets()` to call into `nested_view_from_values_offsets_lengths()`
* Applies @fx.wrap to `nested_view_from_values_offsets_lengths()`. This is now the primary entry-point on the python side to the underlying NJT view construction ATen op. Using `fx.wrap()` avoids problems outlined [here](https://docs.google.com/document/d/1OZahza_hW5SqOMxcTqvMx5eZiR1Hsu8yZboa71JGrIw).
* (side note) Computes min / max for the `torch.nested.as_nested_tensor(t)` case since this is easy (they're both `t.shape[1]`)

TODO: need some `torch.compile` tests representative of FIRST

Differential Revision: [D59638229](https://our.internmc.facebook.com/intern/diff/D59638229)

jbschlosser committed:
Use @fx.wrap for python-side NJT construction

[ghstack-poisoned]

YuqingJ commented:
Thanks for hiding the fx.wrap from users. Could you please import this to internal and then I can run some jobs to verify it works in prod? 
By the way, just confirming, will it be fine if we call fx.wrap on both nested_view_from_values_offsets_lengths and the model side function at the same time (wrap on a wrap)?  If not, might need to check for compatibility too.

jbschlosser commented:
> Could you please import this to internal and then I can run some jobs to verify it works in prod?

Sure thing!

> By the way, just confirming, will it be fine if we call fx.wrap on both nested_view_from_values_offsets_lengths and the model side function at the same time (wrap on a wrap)? If not, might need to check for compatibility too.

My understanding is that the outer wrapped function will end up in the fx graph and the inner wrapped function (i.e. `nested_view_from_values_offsets_lengths()`) won't be. I think things will still function like this, but if we want the BC surface to be `nested_view_from_values_offsets_lengths()` instead of the model-side function, then it would be good to eventually remove the outer fx.wrap from the model-side function.

jbschlosser committed:
Update on "Use @fx.wrap for python-side NJT construction"


This PR:
* Changes `nested_view_from_values_offsets()` to call into `nested_view_from_values_offsets_lengths()`
* Applies fx.wrap to `nested_view_from_values_offsets_lengths()`. This is now the primary entry-point on the python side to the underlying NJT view construction ATen op. Using `fx.wrap()` avoids problems outlined [here](https://docs.google.com/document/d/1OZahza_hW5SqOMxcTqvMx5eZiR1Hsu8yZboa71JGrIw).
* (side note) Computes min / max for the `torch.nested.as_nested_tensor(t)` case since this is easy (they're both `t.shape[1]`)

TODO: need some `torch.compile` tests representative of FIRST

[ghstack-poisoned]

soulitzer commented:
It would be nice if the bc-surface introduced by fx.wrap match the public API, e.g. fx.wrap on `torch.nested.nested_tensor_from_jagged`, but if I understand correctly, the reason we don't do this is because we need to make sure the fx.wrapped function is called in the correct way.

jbschlosser committed:
Update on "Use @fx.wrap for python-side NJT construction"


This PR:
* Changes `nested_view_from_values_offsets()` to call into `nested_view_from_values_offsets_lengths()`
* Applies fx.wrap to `nested_view_from_values_offsets_lengths()`. This is now the primary entry-point on the python side to the underlying NJT view construction ATen op. Using `fx.wrap()` avoids problems outlined [here](https://docs.google.com/document/d/1OZahza_hW5SqOMxcTqvMx5eZiR1Hsu8yZboa71JGrIw).
* (side note) Computes min / max for the `torch.nested.as_nested_tensor(t)` case since this is easy (they're both `t.shape[1]`)

TODO: need some `torch.compile` tests representative of FIRST

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59638229).

YuqingJ commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -508,8 +508,9 @@ def _nt_view_dummy() -> torch.Tensor:
-    values, offsets, ragged_idx=1, min_seqlen=None, max_seqlen=None
+@torch.fx.wrap
+def nested_view_from_values_offsets_lengths(
+    values, offsets, lengths, ragged_idx=1, min_seqlen=None, max_seqlen=None
```
can length has a default None value?

davidberard98 reviewed (commented):
fx usage seems correct to me, though maybe let's wait for @YuqingJ 's tests (and her comment about length=None seems good too?)

also, not sure if you prefer this or not, but another option is to wrap `_nt_view_dummy()`.

and idk how to do this, but it might also be nice to mark this function as needing to be backward compatible so that CI would catch changes to its signature

davidberard98 commented on torch/nested/__init__.py:
```diff
@@ -397,7 +398,7 @@ def nested_tensor_from_jagged(
-    return nested_view_from_values_offsets_lengths(
-        values, offsets, lengths, ragged_idx=jagged_dim, min_seqlen=min_seqlen, max_seqlen=max_seqlen)
+    # NB: the full path -must- be used here instead of importing so that @fx.wrap works
+    return torch.nested._internal.nested_tensor.nested_view_from_values_offsets_lengths(
```
I think we need to also `import torch.nested._internal.nested_tensor`

davidberard98 replied:
okay, apparently we can't do this, repro:

```python
import torch
import torch.nested._internal.nested_tensor

asdf = "asdf"

def fn(x, y):
    return torch.nested.nested_tensor_from_jagged(x, y)

x = torch.randn(10, 16)
y = torch.tensor([0, 1, 3, 6, 10])

fn_fx = torch.fx.symbolic_trace(fn)
torch.fx.symbolic_trace(fn_fx)
```

davidberard98 replied:
i think this is why - fns from torch/ have some special handling here:

https://www.internalfb.com/code/fbsource/[4060bddc90fe1f9357b0a73656e0879678a364d1]/fbcode/caffe2/torch/fx/graph.py?lines=398-400

jbschlosser commented:
Closing this PR according to @davidberard98's findings that this doesn't fix the fx-tracing problems. For now, we'll require that users apply `fx.wrap()` to functions that call `nested_tensor_from_jagged()`.
